File: repos_estudiantes_ALVAREZ_primera_entrega/CARDON_PAZ_LUIS_ABDEL/INTRODUCCION_POO/EJERCICIO_15/Buzon.java
package INTRODUCCION_POO.EJERCICIO_15;

public class Buzon {
    private int nro;
    private int nroC;
    private String[][] cartas;
    private Carta[] detallesCartas;
    
    public Buzon(int nro) {
        this.nro = nro;
        this.nroC = 0;
        this.cartas = new String[50][3];
        this.detallesCartas = new Carta[50];
    }
    
    public void agregarCarta(String codigo, String remitente, String destinatario, String descripcion) {
        if (nroC < 50) {
            cartas[nroC][0] = codigo;
            cartas[nroC][1] = remitente;
            cartas[nroC][2] = destinatario;
            detallesCartas[nroC] = new Carta(codigo, descripcion);
            nroC++;
            System.out.println("Carta " + codigo + " agregada al buzón " + nro);
        } else {
            System.out.println("¡Buzón lleno! No se puede agregar más cartas.");
        }
    }
    
    public boolean eliminarCarta(String codigo) {
        for (int i = 0; i < nroC; i++) {
            if (cartas[i][0] != null && cartas[i][0].equals(codigo)) {
                // Desplazar elementos hacia atrás
                for (int j = i; j < nroC - 1; j++) {
                    cartas[j][0] = cartas[j + 1][0];
                    cartas[j][1] = cartas[j + 1][1];
                    cartas[j][2] = cartas[j + 1][2];
                    detallesCartas[j] = detallesCartas[j + 1];
                }
                cartas[nroC - 1][0] = null;
                cartas[nroC - 1][1] = null;
                cartas[nroC - 1][2] = null;
                detallesCartas[nroC - 1] = null;
                nroC--;
                System.out.println("Carta " + codigo + " eliminada.");
                return true;
            }
        }
        System.out.println("Carta con código " + codigo + " no encontrada.");
        return false;
    }
    
    public int contarCartasPorDestinatario(String destinatario) {
        int count = 0;
        for (int i = 0; i < nroC; i++) {
            if (cartas[i][2] != null && cartas[i][2].equalsIgnoreCase(destinatario)) {
                count++;
            }
        }
        return count;
    }
    
    public String verificarRemitenteRecibio(String remitente) {
        StringBuilder resultado = new StringBuilder();
        boolean encontrado = false;
        
        for (int i = 0; i < nroC; i++) {
            if (cartas[i][2] != null && cartas[i][2].equalsIgnoreCase(remitente)) {
                resultado.append("El remitente '").append(remitente)
                        .append("' recibió una carta de '").append(cartas[i][1]).append("'\n");
                encontrado = true;
            }
        }
        
        if (!encontrado) {
            resultado.append("El remitente '").append(remitente).append("' no ha recibido cartas.");
        }
        
        return resultado.toString();
    }
    
    public void buscarPalabraClave(String palabra) {
        System.out.println("Buscando palabra '" + palabra + "' en las descripciones:");
        boolean encontrada = false;
        
        for (int i = 0; i < nroC; i++) {
            if (detallesCartas[i] != null && detallesCartas[i].contienePalabra(palabra)) {
                System.out.println("Coincidencia encontrada:");
                System.out.println("Código: " + cartas[i][0]);
                System.out.println("Remitente: " + cartas[i][1]);
                System.out.println("Destinatario: " + cartas[i][2]);
                System.out.println("Descripción: " + detallesCartas[i].getDescripcion());
                System.out.println("---");
                encontrada = true;
            }
        }
        
        if (!encontrada) {
            System.out.println("No se encontraron coincidencias para la palabra '" + palabra + "'");
        }
    }
    
    public void mostrarCartas() {
        System.out.println("=== BUZÓN " + nro + " ===");
        System.out.println("Número de cartas: " + nroC);
        
        if (nroC == 0) {
            System.out.println("El buzón está vacío.");
            return;
        }
        
        for (int i = 0; i < nroC; i++) {
            System.out.println("Carta " + (i + 1) + ":");
            System.out.println("  Código: " + cartas[i][0]);
            System.out.println("  Remitente: " + cartas[i][1]);
            System.out.println("  Destinatario: " + cartas[i][2]);
            if (detallesCartas[i] != null) {
                System.out.println("  Descripción: " + detallesCartas[i].getDescripcion());
            }
            System.out.println("---");
        }
    }
    
    public int getNro(){ 
        return nro; 
    }
    public int getNroC(){ 
        return nroC; 
    }
    public String[][] getCartas(){ 
        return cartas; 
    }
}

File: repos_estudiantes_ALVAREZ_primera_entrega/ARUQUIPA_VARGAS_ANA_GLORIA/INTRODUCCION/Ejercicio15/Buzon.py
class Carta:
    def __init__(self, codigo: str, descripcion: str):
        self.codigo = codigo
        self.descripcion = descripcion

    def mostrar(self):
        print(f"Código: {self.codigo}")
        print("Descripción:")
        print(self.descripcion)
        print("-" * 40)


class Buzon:
    def __init__(self, nro: int):
        self.nro = nro
        self.nroC = 0
        self.c = [["" for _ in range(3)] for _ in range(50)] 

    def agregar_carta(self, codigo: str, remitente: str, destinatario: str):
        if self.nroC < 50:
            self.c[self.nroC][0] = codigo
            self.c[self.nroC][1] = remitente
            self.c[self.nroC][2] = destinatario
            self.nroC += 1
        else:
            print("Buzón lleno")

    def eliminar(self, codigo: str):
        for i in range(self.nroC):
            if self.c[i][0] == codigo:
                for j in range(i, self.nroC - 1):
                    self.c[j] = self.c[j + 1]
                self.c[self.nroC - 1] = ["", "", ""]
                self.nroC -= 1
                print(f"Carta con código {codigo} eliminada del buzón {self.nro}.")
                return True
        print(f"No se encontró carta con código {codigo} en buzón {self.nro}.")
        return False

    def contar(self, destinatario: str) -> int:
        return sum(1 for i in range(self.nroC) if self.c[i][2] == destinatario)

    def recibieron(self):
        remitentes = set()
        destinatarios = set()
        for i in range(self.nroC):
            remitentes.add(self.c[i][1])
            destinatarios.add(self.c[i][2])
        return remitentes.intersection(destinatarios)

    def mostrar_cartas(self, cartas_descripciones: dict):
        print(f"Buzón nro {self.nro} con {self.nroC} cartas:")
        for i in range(self.nroC):
            codigo, remitente, destinatario = self.c[i]
            descripcion = cartas_descripciones.get(codigo, "Sin descripción")
            print(f"Código: {codigo}, Remitente: {remitente}, Destinatario: {destinatario}")
            print(f"Descripción:\n{descripcion}")
            print("-" * 40)



